### PR TITLE
New tool: wal-reader

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR adds new `wal-reader` tool to our toolset. WAL reader reads current state of the wal (latest checkpoint + segment files after checkpoint), and prints individual entries from it, as they are found.

#### Checklist

- [na] Tests updated
- [na] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tools/wal-reader/main.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	for i := startFrom; i <= endAt; i++ {
+		log.Println("replaying WAL segment", i)
+		s, err := wlog.OpenReadSegment(wlog.SegmentName(walDir, i))
+		if err != nil {
+			return errors.Wrapf(err, "open WAL segment: %d", i)
+		}
+
+		sr, err := wlog.NewSegmentBufReaderWithOffset(0, s)
+		if errors.Is(err, io.EOF) {
+			// File does not exist.
```
Comment (pracucci) [inline tools/wal-reader/main.go:97]: If the file doesn't exist, isn't the `OpenReadSegment()` to return an error?

File: tools/wal-reader/main.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+			continue
+		}
+		if err != nil {
+			return errors.Wrapf(err, "reader for segment %d", i)
+		}
+		err = printWalEntries(wlog.NewReader(sr), series, printSeriesEntries, printSeriesWithSampleEntries, &minSampleTime, &maxSampleTime)
+		if err := sr.Close(); err != nil {
```
Comment (pracucci) [inline tools/wal-reader/main.go:104]: [nit] I would rename `err` to `closeErr` just for clarity. The different scope makes it a bit tricky to read this code.

File: tools/wal-reader/main.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		log.Println("no checkpoint found, skipping")
+	}
+
+	// Find the last segment.
+	_, endAt, e := wlog.Segments(walDir)
+	if e != nil {
+		return errors.Wrap(e, "finding WAL segments")
+	}
+
+	startFrom++
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply tools/wal-reader/main.go:87]: The reason is that `startFrom` comes from `checkpoint` name:

```
/data/tsdb/redacted # ls -l wal/
total 4867568
-rw-r--r--    1 root     root     134217728 Aug 18 12:10 00385216
-rw-r--r--    1 root     root     134217728 Aug 18 12:12 00385217
-rw-r--r--    1 root     root     134217728 Aug 18 12:15 00385218
[...]
-rw-r--r--    1 root     root     134217728 Aug 18 13:39 00385251
-rw-r--r--    1 root     root     134217728 Aug 18 13:42 00385252
-rw-r--r--    1 root     root     116190840 Aug 18 13:44 00385253
drwxr-xr-x    2 root     root          4096 Aug 18 13:09 checkpoint.00385215
```

But `checkpoint.00385215` already contains segment `00385215` and next segment we need to read is `00385216`.

Also TSDB Head does the same: https://github.com/grafana/mimir-prometheus/blob/70e41fc5ac008d623820dcacdbbcbedbd64ca11f/tsdb/head.go#L713-L730

notice how `startFrom` is incremented on line 719, after loading checkpoint.

I will add comment to the code.

File: tools/wal-reader/main.go (hunk 1 above)
Comment (pstibrany) [reply tools/wal-reader/main.go:97]: I copied the code from https://github.com/grafana/mimir-prometheus/blob/70e41fc5ac008d623820dcacdbbcbedbd64ca11f/tsdb/head.go#L731-L744

but you're right. This comment makes no sense. In fact given that we always use offset 0 in the call, error will never be returned here. I'll replace this with normal error handling here.

File: tools/wal-reader/main.go (hunk 2 above)
Comment (pstibrany) [reply tools/wal-reader/main.go]: Done in https://github.com/grafana/mimir/pull/5780/commits/c879a353c126fc2d550761cf3ddb2864d583667d.

File: tools/wal-reader/main.go (hunk 1 above)
Comment (pstibrany) [reply tools/wal-reader/main.go]: Done in https://github.com/grafana/mimir/pull/5780/commits/c879a353c126fc2d550761cf3ddb2864d583667d

File: tools/wal-reader/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply tools/wal-reader/main.go:88]: Done in https://github.com/grafana/mimir/pull/5780/commits/c879a353c126fc2d550761cf3ddb2864d583667d

**Pull request merged**